pkg/metallb/frr.go

```diff
@@ -0,0 +1,559 @@
+	}
+}
+
+// Exists checks whether the given FrrConfiguration exists.
+func (builder *FrrConfigurationBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof(
+		"Checking if FrrConfiguration %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// WithBGPRouter defines a BGP router as localASN.
+func (builder *FrrConfigurationBuilder) WithBGPRouter(localASN uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP router with an ASN: %v", localASN)
+
+	if strconv.Itoa(int(localASN)) == "" {
+		glog.V(100).Infof("The local ASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'local AS' cannot be empty"
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
+		builder.Definition.Spec.BGP.Routers = []frrtypes.Router{{ASN: localASN}}
+	} else {
+		builder.Definition.Spec.BGP.Routers = append(builder.Definition.Spec.BGP.Routers, frrtypes.Router{ASN: localASN})
+	}
+
+	return builder
+}
+
+// WithToReceiveModeAll allows all incoming routes.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeAll(bgpPeerIP string,
+	remoteAS uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive all routes in namespace %s with mode type: all",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if bgpPeerIP == "" {
+		glog.V(100).Infof("The BGP peer IP should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the bgpPassword %s is an empty string", bgpPeerIP)
+	}
+
+	if strconv.Itoa(int(remoteAS)) == "" {
+		glog.V(100).Infof("The remoteASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'remoteAS' cannot be empty when defining receive all mode"
+	}
+
+	// Defining the neighbor with peer IP and ASN information.
+	if len(builder.Definition.Spec.BGP.Routers[0].Neighbors) == 0 {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors = []frrtypes.Neighbor{{
+			Address:   bgpPeerIP,
+			ASN:       remoteAS,
+			ToReceive: frrtypes.Receive{Allowed: frrtypes.AllowedInPrefixes{Mode: frrtypes.AllowAll}},
+		},
+		}
+	} else {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors = append(builder.Definition.Spec.BGP.Routers[0].Neighbors,
+			frrtypes.Neighbor{
+				Address:   bgpPeerIP,
+				ASN:       remoteAS,
+				ToReceive: frrtypes.Receive{Allowed: frrtypes.AllowedInPrefixes{Mode: frrtypes.AllowAll}},
+			})
+	}
+
+	return builder
+}
+
+// WithToReceiveModeFiltered allows only specified prefixes to be received.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeFiltered(bgpPeerIP string, remoteAS uint32,
```

Comment:
This func should not create a neighbor, just change Toreceive. So it should get neighborIndex and change requested neighbor's Toreceive

Reply:
I think it should also add the prefixes. I will remove the neighbor info and keep the prefixes.